src/cli/util.js

```diff
@@ -397,33 +399,130 @@ function createIgnorerFromContextOrDie(context) {
+  // TODO: use `ignore` option for `globby`
+  const extraPatterns = [
+    // The '!./' globs are due to https://github.com/prettier/prettier/issues/2110
+    ...(withNodeModules ? [] : ["!**/node_modules/**", "!./node_modules/**"]),
+    "!**/.{git,svn,hg}/**",
+    "!./.{git,svn,hg}/**"
+  ];
+
+  const allFilesPatterns = ["**/*", ...extraPatterns];
+  // Ignores files in version control systems directories and `node_modules`
+  const ignoredDirectories = [".git", ".svn", ".hg"];
+  if (!withNodeModules) {
+    ignoredDirectories.push("node_modules");
+  }
+
+  let files = [];
+  const cwd = process.cwd();
+  const patterns = [];
+
+  for (const pattern of maybePatterns) {
+    const absolutePath = path.resolve(cwd, pattern);
+    const stat = statSafeSync(absolutePath);
+    if (
+      stat &&
+      path
+        .relative(process.cwd(), absolutePath)
+        .split(path.sep)
+        .some(directory => ignoredDirectories.includes(directory))
+    ) {
+      continue;
+    }
 
-    if (filePaths.length === 0) {
+    if (
+      stat &&
+      stat.isDirectory() &&
+      // `dot pattern` and `expand directories` support need handle differently
+      // for backward compatibility reason only expand `directories` like a glob pattern
+      // see https://github.com/prettier/prettier/pull/6639#issuecomment-548949954
+      isGlobPattern(pattern)
+    ) {
+      // Don't use `files.push(...)`, there is limitation on the number of arguments.
+      files = [
+        ...files,
+        ...globby.sync(allFilesPatterns, {
+          ...globbyOptions,
+          cwd: absolutePath
+        })
+      ];
+      continue;
+    }
+
+    if (stat && stat.isFile()) {
+      files.push(absolutePath);
+      continue;
+    }
+
+    if (isGlobPattern(pattern)) {
+      patterns.push(pattern);
+    }
+  }
+
+  if (patterns.length > 0) {
+    try {
+      files = [
+        ...files,
+        ...globby.sync([...patterns, ...extraPatterns], globbyOptions)
+      ];
+    } catch (error) {
       context.logger.error(
-        `No matching files. Patterns tried: ${patterns.join(" ")}`
+        `Unable to expand glob patterns: ${[
+          ...maybePatterns,
+          ...extraPatterns
+        ].join(" ")}\n${error.message}`
       );
+      // Don't exit the process if one pattern failed
       process.exitCode = 2;
-      return;
     }
-    filePaths.forEach(filePath => callback(filePath));
-  } catch (error) {
+  }
+
+  if (files.length === 0) {
     context.logger.error(
-      `Unable to expand glob patterns: ${patterns.join(" ")}\n${error.message}`
+      `No matching files. Patterns tried: ${[
+        ...maybePatterns,
+        ...extraPatterns
+      ].join(" ")}`
     );
-    // Don't exit the process if one pattern failed
     process.exitCode = 2;
+    return;
+  }
+  files = uniq(files.map(file => path.relative(process.cwd(), file)))
+    // keeping file orders for backward compatibility
+    .sort((a, b) => a.localeCompare(b));
```

Comment:
Is this compatibility thing important? This sorting prevents us from refactoring this function into a generator, which would be the best solution to the problems of big arrays.

Reply:
@thorn0 It's not important, but makes it stable, it seem required for tests. And this order is not even good, I tried to find a better/stable sort package, but seems `sort-paths` `path-sort` `path-sort2` all not that popular. So I decide to ignore this now.

Reply:
Valid points.

> And this order is not even good

What's wrong with it?

Reply:
```text
a.js
b/a/a.js
b/b.js
c.js
```

I prefer by depth and file first

```text
a.js
c.js
b/b.js
b/a/a.js
```

Like

```js
.sort(
  (a, b) => (a.split('/').length - b.split('/').length) || a.localeCompare(b)
)
```

Reply:
I see, that's not what 1.19.1 does anyway.